### DISCUSSION TITLE
AttributeError when sampling with PyMC model using conditional probabilities

### DISCUSSION THREAD

**Discussion by yk4978** [pymc-devs/pymc3#7618] [Q&A]
I am encountering an error when trying to sample from a PyMC model. In the model, I am defining a set of `Bernoulli` variables with conditional probabilities based on another parent `Bernoulli` variable `A`. The model definition itself works fine, but when I attempt to sample and calculate the posterior, I get the following error:
```AttributeError: 'Scratchpad' object has no attribute 'ufunc'```

Here is the code I am using:

```python
import pymc as pm

n = 40
with pm.Model() as model:
    A = pm.Bernoulli('A', p=0.001)
    for i in range(n):
        if i < n // 2:
            pm.Bernoulli(f"A{i}", p=pm.math.switch(A, 0.6, 0.01), observed=1)
        else:
            pm.Bernoulli(f"A{i}", p=pm.math.switch(A, 0.6, 0.01))
    trace = pm.sample(1000, return_inferencedata=False, cores=1)

probs = sum(trace['A']) / len(trace['A'])
print(f"probs: {probs}")
```
Can anyone help me identify what might be causing this error and suggest a possible solution?

**Comment by muhammadalmaskhan** [comment]
Hi @yk4978, 

The error you are encountering, ``` AttributeError: 'Scratchpad' ``` object has no attribute ``` 'ufunc' ```, is likely caused by an issue with how conditional probabilities are defined within the ``` pm.Bernoulli distributions ( ) ``` in your model. Specifically, the ``` pm.math.switch ``` function, which is used to define the conditional probabilities, is not being evaluated correctly during the sampling process.

In PyMC, when you use pm.math.switch within a Bernoulli distribution to conditionally set the probability parameter p, it may not properly handle the symbolic expressions during the sampling phase, leading to the AttributeError

## Solution 

The issue can be resolved by pre-computing the conditional probability and storing it in a variable outside the pm.Bernoulli call. This ensures that the conditional expression is fully evaluated before it is passed to the distribution.

Here’s an updated version of your code with this fix

``` import pymc as pm

n = 40
with pm.Model() as model:
    A = pm.Bernoulli('A', p=0.001)
    
    # Pre-compute the conditional probability for the Bernoulli distributions
    for i in range(n):
        p_val = pm.math.switch(A, 0.6, 0.01)  # Define the conditional probability
        if i < n // 2:
            pm.Bernoulli(f"A{i}", p=p_val, observed=1)  # Observed for the first half
        else:
            pm.Bernoulli(f"A{i}", p=p_val)  # Not observed for the second half
    
    # Perform sampling
    trace = pm.sample(1000, return_inferencedata=False, cores=1)

# Calculate the posterior probability of A
probs = sum(trace['A']) / len(trace['A'])
print(f"Posterior probability of A: {probs}")
 ```
 

-  The key change is the computation of the conditional probability p_val using pm.math.switch before it is passed to the Bernoulli distribution. By storing the conditional probability in a variable, you ensure that the model is built with properly evaluated values, preventing the AttributeError.

- The pm.math.switch function is still used to set the probability based on the value of A, but the conditional probability is now explicitly calculated in advance.

Conclusion:
This approach ensures that the model's probabilistic expressions are evaluated correctly during the sampling process, and it should resolve the AttributeError you're experiencing. Please try running this updated code, and let me know if the issue persists.

~almas
